OpenCV 4.0.1 OSX build framework fails on xcodebuild

##### System information (version)

- OpenCV => 4.0.1
- Operating System / Platform => Mojave 10.14.2
- Compiler => CMake 3.13.3
Mac mini (Late 2014) cpu 2.6 GHz Intel Core i5

##### Detailed description

When I try to create framework for OSx I have this error:

```
** BUILD FAILED **

The following build commands failed:
	CompileC platforms/osx/res/build/build-x86_64-macosx/modules/world/OpenCV.build/Release/opencv_world.build/Objects-normal/x86_64/stereobm.o modules/calib3d/src/stereobm.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)
============================================================
ERROR: Command '['xcodebuild', 'MACOSX_DEPLOYMENT_TARGET=10.9', 'ARCHS=x86_64', '-sdk', 'macosx', '-configuration', 'Release', '-parallelizeTargets', '-jobs', '4', '-target', 'ALL_BUILD', 'build']' returned non-zero exit status 65
============================================================
Traceback (most recent call last):
  File "/Users/me/Downloads/opencv-master/platforms/ios/build_framework.py", line 117, in build
    self._build(outdir)
  File "/Users/me/Downloads/opencv-master/platforms/ios/build_framework.py", line 109, in _build
    self.buildOne(t[0], t[1], mainBD, cmake_flags)
  File "/Users/me/Downloads/opencv-master/platforms/ios/build_framework.py", line 205, in buildOne
    execute(buildcmd + ["-target", "ALL_BUILD", "build"], cwd = builddir)
  File "/Users/me/Downloads/opencv-master/platforms/ios/build_framework.py", line 39, in execute
    retcode = check_call(cmd, cwd = cwd)
  File "/usr/local/Cellar/python@2/2.7.15_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 190, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['xcodebuild', 'MACOSX_DEPLOYMENT_TARGET=10.9', 'ARCHS=x86_64', '-sdk', 'macosx', '-configuration', 'Release', '-parallelizeTargets', '-jobs', '4', '-target', 'ALL_BUILD', 'build']' returned non-zero exit status 65
```

I"ve also tried MACOSX_DEPLOYMENT_TARGET = 10.14
but got another error:

```
ERROR: Command '['libtool', '-static', '-o', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/lib/Release/libopencv_merged.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/libopencv_world.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibtiff.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibpng.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibwebp.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libquirc.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libippiw.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibprotobuf.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libIlmImf.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libippicv.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libade.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libzlib.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibjasper.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibjpeg-turbo.a']' returned non-zero exit status 1
============================================================
Traceback (most recent call last):
  File "/Users/me/Downloads/opencv-master/platforms/ios/build_framework.py", line 117, in build
    self._build(outdir)
  File "/Users/me/Downloads/opencv-master/platforms/ios/build_framework.py", line 112, in _build
    self.mergeLibs(mainBD)
  File "/Users/me/Downloads/opencv-master/platforms/ios/build_framework.py", line 213, in mergeLibs
    execute(["libtool", "-static", "-o", res] + libs + libs3)
  File "/Users/me/Downloads/opencv-master/platforms/ios/build_framework.py", line 39, in execute
    retcode = check_call(cmd, cwd = cwd)
  File "/usr/local/Cellar/python@2/2.7.15_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 190, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['libtool', '-static', '-o', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/lib/Release/libopencv_merged.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/libopencv_world.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibtiff.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibpng.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibwebp.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libquirc.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libippiw.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibprotobuf.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libIlmImf.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libippicv.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libade.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/libzlib.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibjasper.a', '/Users/me/Downloads/opencv-master/platforms/osx/res/build/build-x86_64-macosx/install/lib/3rdparty/liblibjpeg-turbo.a']' returned non-zero exit status 1

```

I've tried master branch and also get same error but for 10.13 it get finally compiled
and when I dropped framework into osx project I got errors:

```
Undefined symbols for architecture x86_64:
  "_cblas_cgemm", referenced from:
      lapack_gemm32fc(float const*, unsigned long, float const*, unsigned long, float, float const*, unsigned long, float, float*, unsigned long, int, int, int, int) in opencv2(hal_internal.o)
      lapack_gemm64fc(double const*, unsigned long, double const*, unsigned long, double, double const*, unsigned long, double, double*, unsigned long, int, int, int, int) in opencv2(hal_internal.o)
  "_cblas_dgemm", referenced from:
      lapack_gemm64f(double const*, unsigned long, double const*, unsigned long, double, double const*, unsigned long, double, double*, unsigned long, int, int, int, int) in opencv2(hal_internal.o)
  "_cblas_sgemm", referenced from:
      lapack_gemm32f(float const*, unsigned long, float const*, unsigned long, float, float const*, unsigned long, float, float*, unsigned long, int, int, int, int) in opencv2(hal_internal.o)
      lapack_gemm64f(double const*, unsigned long, double const*, unsigned long, double, double const*, unsigned long, double, double*, unsigned long, int, int, int, int) in opencv2(hal_internal.o)
  "_cblas_zgemm", referenced from:
      lapack_gemm64fc(double const*, unsigned long, double const*, unsigned long, double, double const*, unsigned long, double, double*, unsigned long, int, int, int, int) in opencv2(hal_internal.o)
  "_clBuildProgram", referenced from:
      _clBuildProgram_pfn in opencv2(opencl_core.o)
```

##### Steps to reproduce
cd osx
python build_framework.py res

Here error for target 10.9

```
In file included from /Users/me/Downloads/opencv-master/modules/calib3d/src/stereobm.cpp:52:
In file included from /Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin.hpp:149:
/Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin_sse_em.hpp:103:1: error: 
      always_inline function '_mm_cvtepu8_epi16' requires target feature
      'sse4.1', but would be inlined into function '_v128_cvtepu8_epi16' that is
      compiled without support for 'sse4.1'
OPENCV_HAL_SSE_WRAP_1(cvtepu8_epi16, __m128i)
^
/Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin_sse_em.hpp:17:14: note: 
      expanded from macro 'OPENCV_HAL_SSE_WRAP_1'
    { return _mm_##fun(a); }
             ^
<scratch space>:95:1: note: expanded from here
_mm_cvtepu8_epi16
^
In file included from /Users/me/Downloads/opencv-master/modules/calib3d/src/stereobm.cpp:52:
In file included from /Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin.hpp:149:
/Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin_sse_em.hpp:104:1: error: 
      always_inline function '_mm_cvtepi8_epi16' requires target feature
      'sse4.1', but would be inlined into function '_v128_cvtepi8_epi16' that is
      compiled without support for 'sse4.1'
OPENCV_HAL_SSE_WRAP_1(cvtepi8_epi16, __m128i)
^
/Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin_sse_em.hpp:17:14: note: 
      expanded from macro 'OPENCV_HAL_SSE_WRAP_1'
    { return _mm_##fun(a); }
             ^
<scratch space>:97:1: note: expanded from here
_mm_cvtepi8_epi16
^
In file included from /Users/me/Downloads/opencv-master/modules/calib3d/src/stereobm.cpp:52:
In file included from /Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin.hpp:149:
/Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin_sse_em.hpp:108:1: error: 
      always_inline function '_mm_cvtepi16_epi32' requires target feature
      'sse4.1', but would be inlined into function '_v128_cvtepi16_epi32' that
      is compiled without support for 'sse4.1'
OPENCV_HAL_SSE_WRAP_1(cvtepi16_epi32, __m128i)
^
/Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin_sse_em.hpp:17:14: note: 
      expanded from macro 'OPENCV_HAL_SSE_WRAP_1'
    { return _mm_##fun(a); }
             ^
<scratch space>:105:1: note: expanded from here
_mm_cvtepi16_epi32
^
In file included from /Users/me/Downloads/opencv-master/modules/calib3d/src/stereobm.cpp:52:
In file included from /Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin.hpp:149:
/Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin_sse_em.hpp:107:1: error: 
      always_inline function '_mm_cvtepu16_epi32' requires target feature
      'sse4.1', but would be inlined into function '_v128_cvtepu16_epi32' that
      is compiled without support for 'sse4.1'
OPENCV_HAL_SSE_WRAP_1(cvtepu16_epi32, __m128i)
^
/Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin_sse_em.hpp:17:14: note: 
      expanded from macro 'OPENCV_HAL_SSE_WRAP_1'
    { return _mm_##fun(a); }
             ^
<scratch space>:103:1: note: expanded from here
_mm_cvtepu16_epi32
^
In file included from /Users/me/Downloads/opencv-master/modules/calib3d/src/stereobm.cpp:52:
In file included from /Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin.hpp:150:
/Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin_sse.hpp:942:22: error: 
      always_inline function '_mm_min_epi32' requires target feature 'sse4.1',
      but would be inlined into function 'v_min' that is compiled without
      support for 'sse4.1'
    return v_int32x4(_mm_min_epi32(a.val, b.val));
                     ^
/Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin_sse.hpp:1582:1: error: 
      always_inline function '_mm_blendv_ps' requires target feature 'sse4.1',
      but would be inlined into function 'v_select' that is compiled without
      support for 'sse4.1'
OPENCV_HAL_IMPL_SSE_SELECT(v_int32x4, _mm_castps_si128, _mm_castsi128_ps, ps)
^
/Users/me/Downloads/opencv-master/modules/core/include/opencv2/core/hal/intrin_sse.hpp:1574:28: note: 
      expanded from macro 'OPENCV_HAL_IMPL_SSE_SELECT'
    return _Tpvec(cast_ret(_mm_blendv_##suffix(cast(b.val), cast(a.val)...
                           ^
<scratch space>:15:1: note: expanded from here
_mm_blendv_ps
^
6 errors generated.
```

--- a/platforms/ios/build_framework.py
+++ b/platforms/ios/build_framework.py
@@ -49,7 +49,7 @@ def getXCodeMajor():
         raise Exception("Failed to parse Xcode version")
 
 class Builder:
-    def __init__(self, opencv, contrib, dynamic, bitcodedisabled, exclude, enablenonfree, targets):
+    def __init__(self, opencv, contrib, dynamic, bitcodedisabled, exclude, enablenonfree, targets, debug):
         self.opencv = os.path.abspath(opencv)
         self.contrib = None
         if contrib:
@@ -63,6 +63,7 @@ class Builder:
         self.exclude = exclude
         self.enablenonfree = enablenonfree
         self.targets = targets
+        self.debug = debug
 
     def getBD(self, parent, t):
 
@@ -125,6 +126,9 @@ class Builder:
     def getToolchain(self, arch, target):
         return None
 
+    def getConfiguration(self):
+        return "Debug" if self.debug else "Release"
+
     def getCMakeArgs(self, arch, target):
 
         args = [
@@ -132,7 +136,7 @@ class Builder:
             "-GXcode",
             "-DAPPLE_FRAMEWORK=ON",
             "-DCMAKE_INSTALL_PREFIX=install",
-            "-DCMAKE_BUILD_TYPE=Release",
+            "-DCMAKE_BUILD_TYPE=%s" % self.getConfiguration(),
         ] + ([
             "-DBUILD_SHARED_LIBS=ON",
             "-DCMAKE_MACOSX_BUNDLE=ON",
@@ -174,7 +178,7 @@ class Builder:
 
         buildcmd += [
                 "-sdk", target.lower(),
-                "-configuration", "Release",
+                "-configuration", self.getConfiguration(),
                 "-parallelizeTargets",
                 "-jobs", str(multiprocessing.cpu_count()),
             ] + (["-target","ALL_BUILD"] if self.dynamic else [])
@@ -201,10 +205,10 @@ class Builder:
             shutil.rmtree(clean_dir)
         buildcmd = self.getBuildCommand(arch, target)
         execute(buildcmd + ["-target", "ALL_BUILD", "build"], cwd = builddir)
-        execute(["cmake", "-P", "cmake_install.cmake"], cwd = builddir)
+        execute(["cmake", "-DBUILD_TYPE=%s" % self.getConfiguration(), "-P", "cmake_install.cmake"], cwd = builddir)
 
     def mergeLibs(self, builddir):
-        res = os.path.join(builddir, "lib", "Release", "libopencv_merged.a")
+        res = os.path.join(builddir, "lib", self.getConfiguration(), "libopencv_merged.a")
         libs = glob.glob(os.path.join(builddir, "install", "lib", "*.a"))
         libs3 = glob.glob(os.path.join(builddir, "install", "share", "OpenCV", "3rdparty", "lib", "*.a"))
         print("Merging libraries:\n\t%s" % "\n\t".join(libs + libs3), file=sys.stderr)
@@ -230,7 +234,7 @@ class Builder:
         shutil.copytree(os.path.join(builddirs[0], "install", "include", "opencv2"), os.path.join(dstdir, "Headers"))
 
         # make universal static lib
-        libs = [os.path.join(d, "lib", "Release", libname) for d in builddirs]
+        libs = [os.path.join(d, "lib", self.getConfiguration(), libname) for d in builddirs]
         lipocmd = ["lipo", "-create"]
         lipocmd.extend(libs)
         lipocmd.extend(["-o", os.path.join(dstdir, name)])
@@ -288,6 +292,7 @@ if __name__ == "__main__":
     parser.add_argument('--iphoneos_archs', default='armv7,armv7s,arm64', help='select iPhoneOS target ARCHS')
     parser.add_argument('--iphonesimulator_archs', default='i386,x86_64', help='select iPhoneSimulator target ARCHS')
     parser.add_argument('--enable_nonfree', default=False, dest='enablenonfree', action='store_true', help='enable non-free modules (disabled by default)')
+    parser.add_argument('--debug', default=False, dest='debug', action='store_true', help='Build "Debug" binaries (disabled by default)')
     args = parser.parse_args()
 
     os.environ['IPHONEOS_DEPLOYMENT_TARGET'] = args.iphoneos_deployment_target
@@ -304,5 +309,5 @@ if __name__ == "__main__":
         [
             (iphoneos_archs, "iPhoneOS"),
             (iphonesimulator_archs, "iPhoneSimulator"),
-        ])
+        ], args.debug)
     b.build(args.out)

--- a/platforms/ios/build_framework.py
+++ b/platforms/ios/build_framework.py
@@ -49,7 +49,7 @@ def getXCodeMajor():
         raise Exception("Failed to parse Xcode version")
 
 class Builder:
-    def __init__(self, opencv, contrib, dynamic, bitcodedisabled, exclude, enablenonfree, targets, debug):
+    def __init__(self, opencv, contrib, dynamic, bitcodedisabled, exclude, enablenonfree, targets, debug, debug_info):
         self.opencv = os.path.abspath(opencv)
         self.contrib = None
         if contrib:
@@ -64,6 +64,7 @@ class Builder:
         self.enablenonfree = enablenonfree
         self.targets = targets
         self.debug = debug
+        self.debug_info = debug_info
 
     def getBD(self, parent, t):
 
@@ -143,7 +144,9 @@ class Builder:
             "-DCMAKE_XCODE_ATTRIBUTE_CODE_SIGNING_REQUIRED=NO",
         ] if self.dynamic else []) + ([
             "-DOPENCV_ENABLE_NONFREE=ON"
-        ] if self.enablenonfree else [])
+        ] if self.enablenonfree else []) + ([
+            "-DBUILD_WITH_DEBUG_INFO=ON"
+        ] if self.debug_info else [])
 
         if len(self.exclude) > 0:
             args += ["-DBUILD_opencv_world=OFF"] if not self.dynamic else []
@@ -293,6 +296,7 @@ if __name__ == "__main__":
     parser.add_argument('--iphonesimulator_archs', default='i386,x86_64', help='select iPhoneSimulator target ARCHS')
     parser.add_argument('--enable_nonfree', default=False, dest='enablenonfree', action='store_true', help='enable non-free modules (disabled by default)')
     parser.add_argument('--debug', default=False, dest='debug', action='store_true', help='Build "Debug" binaries (disabled by default)')
+    parser.add_argument('--debug_info', default=False, dest='debug_info', action='store_true', help='Build with debug information (useful for Release mode: BUILD_WITH_DEBUG_INFO=ON)')
     args = parser.parse_args()
 
     os.environ['IPHONEOS_DEPLOYMENT_TARGET'] = args.iphoneos_deployment_target
@@ -309,5 +313,5 @@ if __name__ == "__main__":
         [
             (iphoneos_archs, "iPhoneOS"),
             (iphonesimulator_archs, "iPhoneSimulator"),
-        ], args.debug)
+        ], args.debug, args.debug_info)
     b.build(args.out)

--- a/platforms/osx/build_framework.py
+++ b/platforms/osx/build_framework.py
@@ -18,10 +18,10 @@ class OSXBuilder(Builder):
     def getBuildCommand(self, archs, target):
         buildcmd = [
             "xcodebuild",
-            "MACOSX_DEPLOYMENT_TARGET=10.9",
+            "MACOSX_DEPLOYMENT_TARGET=10.12",
             "ARCHS=%s" % archs[0],
             "-sdk", target.lower(),
-            "-configuration", "Release",
+            "-configuration", "Debug" if self.debug else "Release",
             "-parallelizeTargets",
             "-jobs", str(multiprocessing.cpu_count())
         ]
@@ -39,10 +39,12 @@ if __name__ == "__main__":
     parser.add_argument('--contrib', metavar='DIR', default=None, help='folder with opencv_contrib repository (default is "None" - build only main framework)')
     parser.add_argument('--without', metavar='MODULE', default=[], action='append', help='OpenCV modules to exclude from the framework')
     parser.add_argument('--enable_nonfree', default=False, dest='enablenonfree', action='store_true', help='enable non-free modules (disabled by default)')
+    parser.add_argument('--debug', action="store_true", help="Build 'Debug' binaries (CMAKE_BUILD_TYPE=Debug)")
+
     args = parser.parse_args()
 
     b = OSXBuilder(args.opencv, args.contrib, False, False, args.without, args.enablenonfree,
         [
             (["x86_64"], "MacOSX")
-        ])
+        ], args.debug)
     b.build(args.out)

--- a/platforms/osx/build_framework.py
+++ b/platforms/osx/build_framework.py
@@ -42,7 +42,8 @@ if __name__ == "__main__":
     parser.add_argument('--without', metavar='MODULE', default=[], action='append', help='OpenCV modules to exclude from the framework')
     parser.add_argument('--enable_nonfree', default=False, dest='enablenonfree', action='store_true', help='enable non-free modules (disabled by default)')
     parser.add_argument('--macosx_deployment_target', default=os.environ.get('MACOSX_DEPLOYMENT_TARGET', MACOSX_DEPLOYMENT_TARGET), help='specify MACOSX_DEPLOYMENT_TARGET')
-    parser.add_argument('--debug', action="store_true", help="Build 'Debug' binaries (CMAKE_BUILD_TYPE=Debug)")
+    parser.add_argument('--debug', action='store_true', help='Build "Debug" binaries (CMAKE_BUILD_TYPE=Debug)')
+    parser.add_argument('--debug_info', action='store_true', help='Build with debug information (useful for Release mode: BUILD_WITH_DEBUG_INFO=ON)')
 
     args = parser.parse_args()
 
@@ -52,5 +53,5 @@ if __name__ == "__main__":
     b = OSXBuilder(args.opencv, args.contrib, False, False, args.without, args.enablenonfree,
         [
             (["x86_64"], "MacOSX")
-        ], args.debug)
+        ], args.debug, args.debug_info)
     b.build(args.out)

--- a/platforms/osx/build_framework.py
+++ b/platforms/osx/build_framework.py
@@ -10,6 +10,8 @@ import os, os.path, sys, argparse, traceback, multiprocessing
 sys.path.insert(0, os.path.abspath(os.path.abspath(os.path.dirname(__file__))+'/../ios'))
 from build_framework import Builder
 
+MACOSX_DEPLOYMENT_TARGET='10.12'  # default, can be changed via command line options or environment variable
+
 class OSXBuilder(Builder):
 
     def getToolchain(self, arch, target):
@@ -18,7 +20,7 @@ class OSXBuilder(Builder):
     def getBuildCommand(self, archs, target):
         buildcmd = [
             "xcodebuild",
-            "MACOSX_DEPLOYMENT_TARGET=10.12",
+            "MACOSX_DEPLOYMENT_TARGET=" + os.environ['MACOSX_DEPLOYMENT_TARGET'],
             "ARCHS=%s" % archs[0],
             "-sdk", target.lower(),
             "-configuration", "Debug" if self.debug else "Release",
@@ -39,9 +41,13 @@ if __name__ == "__main__":
     parser.add_argument('--contrib', metavar='DIR', default=None, help='folder with opencv_contrib repository (default is "None" - build only main framework)')
     parser.add_argument('--without', metavar='MODULE', default=[], action='append', help='OpenCV modules to exclude from the framework')
     parser.add_argument('--enable_nonfree', default=False, dest='enablenonfree', action='store_true', help='enable non-free modules (disabled by default)')
+    parser.add_argument('--macosx_deployment_target', default=os.environ.get('MACOSX_DEPLOYMENT_TARGET', MACOSX_DEPLOYMENT_TARGET), help='specify MACOSX_DEPLOYMENT_TARGET')
     parser.add_argument('--debug', action="store_true", help="Build 'Debug' binaries (CMAKE_BUILD_TYPE=Debug)")
 
     args = parser.parse_args()
+
+    os.environ['MACOSX_DEPLOYMENT_TARGET'] = args.macosx_deployment_target
+    print('Using MACOSX_DEPLOYMENT_TARGET=' + os.environ['MACOSX_DEPLOYMENT_TARGET'])
 
     b = OSXBuilder(args.opencv, args.contrib, False, False, args.without, args.enablenonfree,
         [

